crates/bevy_render/src/mesh/allocator.rs

```diff
@@ -0,0 +1,1025 @@
+    /// bytes, so this is the most convenient unit in this case.
+    pub range: Range<u32>,
+}
+
+/// The index of a single slab.
+#[derive(Clone, Copy, Default, PartialEq, Eq, Hash, Debug)]
+#[repr(transparent)]
+struct SlabId(u32);
+
+/// Data for a single slab.
+#[allow(clippy::large_enum_variant)]
+enum Slab {
+    /// A slab that can contain multiple objects.
+    General(GeneralSlab),
+    /// A slab that contains a single object.
+    LargeObject(LargeObjectSlab),
+}
+
+/// A resizable slab that can contain multiple objects.
+///
+/// This is the normal type of slab used for objects that are below the
+/// [`MeshAllocatorSettings::large_threshold`]. Slabs are divided into *slots*,
+/// which are described in detail in the [`ElementLayout`] documentation.
+struct GeneralSlab {
+    /// The [`Allocator`] thatmanages the objects in this slab.
+    allocator: Allocator,
+
+    /// The GPU buffer that backs this slab.
+    ///
+    /// This may be `None` if the buffer hasn't been created yet. We delay
+    /// creation of buffers until allocating all the meshes for a single frame,
+    /// so that we don't needlessly create and resize buffers when many meshes
+    /// load all at once.
+    buffer: Option<Buffer>,
+
+    /// Allocations that are on the GPU.
+    ///
+    /// The range is in slots.
+    resident_allocations: HashMap<AssetId<Mesh>, SlabAllocation>,
+
+    /// Allocations that are waiting to be uploaded to the GPU.
+    ///
+    /// The range is in slots.
+    pending_allocations: HashMap<AssetId<Mesh>, SlabAllocation>,
+
+    /// The layout of a single element (vertex or index).
+    element_layout: ElementLayout,
+
+    /// The size of this slab in slots.
+    slot_capacity: u32,
+}
+
+/// A slab that contains a single object.
+///
+/// Typically, this is for objects that exceed the
+/// [`MeshAllocatorSettings::large_threshold`]. This is also for objects that
+/// would ordinarily receive their own slab but can't because of platform
+/// limitations, most notably vertex arrays on WebGL 2.
+struct LargeObjectSlab {
+    /// The GPU buffer that backs this slab.
+    ///
+    /// This may be `None` if the buffer hasn't been created yet.
+    buffer: Option<Buffer>,
+
+    /// The layout of a single element (vertex or index).
+    element_layout: ElementLayout,
+}
+
+/// The type of element that a slab can store.
+#[derive(Clone, Copy, PartialEq, Eq, Hash)]
+enum ElementClass {
+    /// Data for a vertex.
+    Vertex,
+    /// A vertex index.
+    Index,
+}
+
+/// Information about the size of individual elements (vertices or indices)
+/// within a slab.
+///
+/// Slab objects are allocated in units of *slots*. Usually, each element takes
+/// up one slot, and so elements and slots are equivalent. Occasionally,
+/// however, a slot may consist of 2 or even 4 elements. This occurs when the
+/// size of an element isn't divisible by [`COPY_BUFFER_ALIGNMENT`] (4). When we
```

Comment:
```suggestion
/// size of an element isn't divisible by [`COPY_BUFFER_ALIGNMENT`]. When we
```

Specifying 4 makes reading the sentence more confusing that not, imo.

Reply:
Removed